dhall-docs/src/Dhall/Docs/Core.hs

```diff
@@ -244,69 +280,133 @@ saveHtml inputAbsDir outputAbsDir packageName DhallFile {path = absFile, ..} = d
     2. @\/a\/c\/index.html@ that will list the @\/a\/c\/b.txt@ file
 
 -}
-createIndexes
-    :: Path Abs Dir                 -- ^ Directory where index.html file will be saved. Used
-                                    --   to link the css resources. It should be a prefix for
-                                    --   each @Path Abs File@ on the second argument
-    -> [(Path Abs File, DhallFile)] -- ^ Html files generated by the tool
-    -> Text                         -- ^ Package name
-    -> IO ()
-createIndexes outputPath htmlFiles packageName = do
-    let toMap file = Map.singleton (Path.parent $ fst file) [file]
-    let filesGroupedByDir = Map.unionsWith (<>) $ map toMap htmlFiles
-
-    let listDirRel dir = do
-            dirs <- fst <$> Path.IO.listDir dir
-            mapM (Path.stripProperPrefix dir) dirs
-
-    let createIndex :: Path Abs Dir -> [(Path Abs File, DhallFile)] -> IO ()
-        createIndex index files = do
-            indexFile <- Path.fromAbsFile . (index </>) <$> Path.parseRelFile "index.html"
-            indexTitle <-
-                if outputPath == index then return $(Path.mkRelDir ".")
-                else Path.stripProperPrefix outputPath index
-            indexList <- Control.Monad.forM files $ \(fileOutputPath, dhallFile) -> do
-                path <- Path.filename <$> Path.stripProperPrefix outputPath fileOutputPath
-                return (path, mType dhallFile)
-            dirList <- listDirRel index
-
-            let relativeResourcesPath = resolveRelativePath outputPath index
-            Lucid.renderToFile indexFile $
-                indexToHtml
-                    indexTitle
-                    indexList
-                    dirList
-                    DocParams { relativeResourcesPath, packageName }
-
-    _ <- Map.traverseWithKey createIndex filesGroupedByDir
-    return ()
-
--- | Generate all of the docs for a package
+createIndexes :: Text -> [DhallFile] -> [(Path Rel File, Text)]
+createIndexes packageName files = map toIndex dirToDirsAndFilesMapAssocs
```

Comment:
The code inside this function feels kind of messy and impenetrable. I'm not sure how to best fix it though.

I somewhat wonder whether the `(Map (Path Rel Dir))` is a good fit for this logic, or if possibly a custom tree structure would be a better fit.

Reply:
IKR!?

The most messy part IMO is the `dirToDirsMap`, that was the best I could do to get it right.

If you're wondering what it does (probably could guide me to another solution) is to try to _compact_ listed paths as much as it can.

for isntance, the folders at the `tasty/data/package` are the following:

```
./
./a/
./a/b/
./a/b/c/
./deep/
./deep/nested/
./deep/nested/folder
```

but on `./deep/` and `./deep/nested/` folders there is no file.

So it's kind of dissapointing if the package directory at `./` listed `./deep/index.html` and the last listed `./deep/nested/index.html` to finally list `./deep/nested/folder/index.html`.

So what it does is to compact paths as much as it cans, without altering the actual directory structure.

An expected output for that directory structure is the output in the golden files:

```
./
./a/
./a/b/
./a/b/c/
./deep/nested/folder
```

I'm still thinking a better solution or at least a more readable one. Get back to you if I find out something.

Reply:
@sjakobi: The most I could do was to use `foldl` instead of passing the whole list to `go` at `dirToDirsMap`.

Another idea I got was the following (pseudocode):

```
def f(sortedDirs):
    stack = [sortedDirs[0]]
    map = { sortedDirs[0] = [ ] }

    for i in range(sortedDirs):
        t = stack.top()
        d = sortedDirs[i]
        if t prefix of d:
            stack.push(d)
            map[t].add(d)
        else:
            while stack.top() not prefix of d:
                stack.pop()
            map[stack.top()].add(d)
            stack.push(d)

    return map
```

... but couldn't write a legible implementation, at least not better than our current one.

I haven't tried yet using a tree-like structure, but I think for that in another PR.